tasks/test.py

```diff
@@ -92,11 +121,128 @@ def install_tools(ctx):
+        print(f"----- Module '{module.full_path()}'")
+        if not module.condition():
+            print("----- Skipped")
+            continue
+
+        with ctx.cd(module.full_path()):
+            vet(ctx, targets=module.targets, rtloader_root=rtloader_root, build_tags=build_tags, arch=arch)
+            staticcheck(ctx, targets=module.targets, build_tags=build_tags, arch=arch)
+
+    # For now we only run golangci_lint on Unix as the Windows env needs more work
+    if sys.platform != 'win32':
+        print(f"--- Flavor {flavor.name}: golangci_lint")
+        for module in modules:
+            print(f"----- Module '{module.full_path()}'")
+            if not module.condition():
+                print("----- Skipped")
+                continue
+
+            with ctx.cd(module.full_path()):
+                golangci_lint(
+                    ctx, targets=module.targets, rtloader_root=rtloader_root, build_tags=build_tags, arch=arch
+                )
+
+
+def test_flavor(
+    ctx,
+    flavor: AgentFlavor,
+    build_tags: List[str],
+    modules: List[GoModule],
+    cmd: str,
+    env: Dict[str, str],
+    args: Dict[str, str],
+    junit_tar: str,
+    save_result_json: str,
+    test_profiler: TestProfiler,
+):
+    """
+    Runs unit tests for given flavor, build tags, and modules.
+    """
+    print(f"--- Flavor {flavor.name}: unit tests")
+
+    failed_modules = []
+    junit_files = []
+
+    args["go_build_tags"] = " ".join(build_tags + ["test"])
+
+    junit_file_flag = ""
+    junit_file = f"junit-out-{flavor.name}.xml"
+    if junit_tar:
+        junit_file_flag = "--junitfile " + junit_file
+    args["junit_file_flag"] = junit_file_flag
+
+    for module in modules:
+        print(f"----- Module '{module.full_path()}'")
+        if not module.condition():
+            print("----- Skipped")
+            continue
+
+        with ctx.cd(module.full_path()):
+            res = ctx.run(
+                cmd.format(
+                    packages=' '.join(f"{t}/..." if not t.endswith("/...") else t for t in module.targets), **args
+                ),
+                env=env,
+                out_stream=test_profiler,
+                warn=True,
+            )
+
+        if res.exited is None or res.exited > 0:
+            failed_modules.append(module.full_path())
+
+        if save_result_json:
+            with open(save_result_json, 'ab') as json_file, open(
+                os.path.join(module.full_path(), GO_TEST_RESULT_TMP_JSON), 'rb'
+            ) as module_file:
+                json_file.write(module_file.read())
+
+        junit_file_path = None
+        if junit_tar:
+            junit_file_path = os.path.join(module.full_path(), junit_file)
+            add_flavor_to_junitxml(junit_file_path, flavor)
+
+        junit_files.append(junit_file_path)
+
+    return junit_files, failed_modules
+
+
+@task(iterable=['flavor'])
 def test(
     ctx,
     module=None,
     targets=None,
+    flavor=None,
```

Comment:
This is an iterable, so maybe better to call it `flavors`? Also, do we really need it to be an iterable? Maybe it's better to accept just a single flavor and call the test function repeatedly. It is slightly less eficient but makes the function simpler

Reply:
If you call it `flavors`, then in the CLI you need to pass `invoke test --flavors iot --flavors base`, which also feels slightly weird. I guess we'd rather have clearer code than clearer CLI args (especially for something that's almost exclusively used in CI), so I can rename it to `flavors`.

I still like having the ability to pass multiple flavors at once, as that way we don't call multiple times the same flavor-independent operations. I feel like it doesn't add that much complexity (calls to functions become loops of calls to functions).

Reply:
Okay, agree that clearer code is more important than clearer CLI, since this is CI-only